### PR TITLE
pkg/fqdn TTL resolver

### PR DESCRIPTION
**Pull request by raybejjani**
`DNSLookupDefaultResolver` now returns TTLs of lookups. This allows supporting low-TTL toFQDN matchNames when rolling over the allowed IPs is desired. 
Unfortunately, the stdlib doesn't expose lookup TTLs and I had to introduce a new dependency and a bunch of new code :( I tried reusing the `Forward` and `Upstream` types in coredns (assuming that they would do some of this logic) but that was more involved. The logic here is relatively simple but difficult to read. To lookup a name:
- Do a lookup for A and AAAA records
- Try each configured server, in turn, when we have an error (a response from the server isn't an error, however)

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5505)

### CODE REVIEW

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

File: pkg/fqdn/lookup.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+	DNSIPs = make(map[string]*DNSIPRecords)
+	DNSErrors = make(map[string]error)
+
+	// This is the top-level list of names to query
+	for _, dnsName := range dnsNames {
+		responseData := &DNSIPRecords{TTL: math.MaxInt32}
+
+		// Query for A & AAAA for each dnsName
+	perTypeToQuery:
+		for _, dnsType := range []uint16{dns.TypeA, dns.TypeAAAA} { // the dns library doesn't use dns.Type
```
Comment (ianvernon) [inline pkg/fqdn/lookup.go]: Why not make this its own slice, `dnsTypes`?

File: pkg/fqdn/lookup.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+			// Try the servers in the order they were configured in resov.conf
```
Comment (ianvernon) [inline pkg/fqdn/lookup.go]: resolv.conf

File: pkg/fqdn/lookup.go (hunk 1 above)
Comment (raybejjani) [reply pkg/fqdn/lookup.go]: I switched to using the dns package's Type type. Introducing a type just to create this slice seems more likely to create confusion than not. Would it be apparent that it's just a slice of things? I would expect custom collection types to represent some complex thing vs `[]uint16`. Still, if you still find it suboptimal I can put in `dnsTypes`.

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Comment by raybejjani** [comment]
test-me-please

**Review by tgraf — APPROVED** [review]

**Pull request merged**